off-chain/src/Network/Types.purs

```diff
@@ -70,42 +125,146 @@ data StatusResponse
+          }
+
+instance decodeAesonActionStatus :: DecodeAeson ActionStatus where
+  decodeAeson s = do
+    obj <- decodeAeson s
+    tag <- getField obj "tag"
+    contents <- getField obj "contents"
+    case tag of
+      "AskForSignature" -> AskForSignature <$> undefined -- FIXME
+      "ToBeProcessed" -> ToBeProcessed <$> decodeAeson contents
+      "ToBeSubmited" -> ToBeSubmited <$> decodeAeson contents
+      "Processing" -> Right Processing
+      "RejectedAtChainBuilder" -> RejectedAtChainBuilder <$> decodeAeson
+        contents
+      "RequireNewSignature" -> Right RequireNewSignature
+      "SubmitError" -> SubmitError <$> decodeAeson contents
+      "NotFound" -> Right NotFound
+      other -> Left
+        (TypeMismatch $ "IncludeActionError: unexpected constructor " <> other)
+
+data GetStatusError = GSOtherError String
+
+derive instance Generic GetStatusError _
+
+instance showStatusResponseError :: Show GetStatusError where
+  show = genericShow
+
+instance encodeAesonStatusResponseError :: EncodeAeson GetStatusError where
+  encodeAeson = case _ of
+    GSOtherError err -> encodeTagged' "GSOtherError" err
+
+instance decodeAesonStatusResponseError :: DecodeAeson GetStatusError where
+  decodeAeson s = do
+    obj <- decodeAeson s
+    tag <- getField obj "tag"
+    contents <- getField obj "contents"
+    case tag of
+      "GSOtherError" -> GSOtherError <$> decodeAeson contents
+      other -> Left
+        (TypeMismatch $ "IncludeActionError: unexpected constructor " <> other)
+
 newtype SendSignedTransaction = SendSignedTransaction
-  { controlNumber :: UUID
+  { uuid :: UUID
   , transaction :: SignedTransaction
   }
 
--- I suspect that this `a` is a type like :
--- `UUID /\ UserAction a`
-newtype LeaderState a = LeaderState
+derive instance Newtype SendSignedTransaction _
+derive newtype instance Show SendSignedTransaction
+
+type LeaderStateInner a =
   { pendingActionsRequest :: Ref $ OrderedMap UUID (UserAction a)
   -- For actions that were part of a previous built chain 
   -- and were removed since a previous action failed.
   , prioritaryPendingActions :: Ref $ OrderedMap UUID (UserAction a)
-  , signatureResponses :: Ref $ OrderedMap UUID (UserAction a)
+  , processing :: Ref $ OrderedMap UUID (UserAction a)
+  , rejectedByChainBuilder :: Ref $ OrderedMap UUID (UserAction a)
+  , waitingForSignature :: Ref $ OrderedMap UUID Transaction
+  , waitingForSubmission :: Ref $ OrderedMap UUID Transaction
+  , errorAtSubmission :: Ref $ OrderedMap UUID Transaction
   , stage :: LeaderServerStage
-  -- We really need to think if we really want this,
-  -- see `UUID` comment.
-  , numberOfActionsRequestsMade :: MutableInt
   }
 
+newtype LeaderState a = LeaderState (LeaderStateInner a)
+
+getFromRefAtLeaderState
+  :: forall a b. LeaderNode a -> (LeaderStateInner a -> Ref b) -> Aff b
+getFromRefAtLeaderState (LeaderNode node) f =
+  liftEffect $ Ref.read $ f (unwrap node.state)
+
+takeFromPending
```

Comment:
as with `getFromRefAtLeaderState`  we can write a more generic function for this for ease of use with all the `OrderedMaps`

Reply:
I've added "setter" of the same shape.